File: app/models/otp.py
from datetime import datetime, timezone
from beanie import Document, Indexed
from pydantic import Field
from app.schemas.auth_otp import OTPPurpose


class OTP(Document):
    """OTP MongoDB Document model for registration verification and passwordless login."""

    email: Indexed(str)
    code: str
    purpose: OTPPurpose
    expires_at: datetime
    is_used: bool = False
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )

    class Settings:
        name = "otps"

    @property
    def is_expired(self) -> bool:
        """Check if OTP has expired."""
        now = datetime.now(timezone.utc)
        exp = self.expires_at
        if exp.tzinfo is None:
            exp = exp.replace(tzinfo=timezone.utc)
        return now > exp

File: src/models/OTP.ts
import mongoose, { Schema, Document } from 'mongoose';

export enum OTPPurpose {
  REGISTRATION = 'registration',
  LOGIN = 'login',
  PASSWORD_RESET = 'password_reset',
}

export interface IOTP extends Document {
  email: string;
  code: string;
  purpose: OTPPurpose;
  expires_at: Date;
  is_used: boolean;
  created_at: Date;
  is_expired: boolean;
}

const OTPSchema = new Schema<IOTP>(
  {
    email: { type: String, required: true, index: true, lowercase: true, trim: true },
    code: { type: String, required: true },
    purpose: { type: String, enum: Object.values(OTPPurpose), required: true },
    expires_at: { type: Date, required: true },
    is_used: { type: Boolean, default: false },
    created_at: { type: Date, default: Date.now },
  }
);

OTPSchema.index({ email: 1, purpose: 1, is_used: 1 });

OTPSchema.virtual('is_expired').get(function () {
  return new Date() > this.expires_at;
});

export const OTP = mongoose.model<IOTP>('OTP', OTPSchema, 'otps');
